test/integration/apiserver/flowcontrol/concurrency_test.go

```diff
@@ -0,0 +1,317 @@
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+
+	flowcontrolv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/master"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	sharedConcurrencyMetricsName      = "apiserver_flowcontrol_request_concurrency_limit"
+	dispatchedRequestCountMetricsName = "apiserver_flowcontrol_dispatched_requests_total"
+	labelPriorityLevel                = "priorityLevel"
+	timeout                           = time.Second * 10
+)
+
+func setup(t testing.TB) (*httptest.Server, *rest.Config, framework.CloseFunc) {
+	opts := framework.MasterConfigOptions{EtcdOptions: framework.DefaultEtcdOptions()}
+	opts.EtcdOptions.DefaultStorageMediaType = "application/vnd.kubernetes.protobuf"
+	masterConfig := framework.NewIntegrationTestMasterConfigWithOptions(&opts)
+	resourceConfig := master.DefaultAPIResourceConfigSource()
+	resourceConfig.EnableVersions(schema.GroupVersion{
+		Group:   "flowcontrol.apiserver.k8s.io",
+		Version: "v1alpha1",
+	})
+	masterConfig.GenericConfig.MaxRequestsInFlight = 5
+	masterConfig.GenericConfig.MaxMutatingRequestsInFlight = 5
+	masterConfig.GenericConfig.OpenAPIConfig = framework.DefaultOpenAPIConfig()
+	masterConfig.ExtraConfig.APIResourceConfigSource = resourceConfig
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	return s, masterConfig.GenericConfig.LoopbackClientConfig, closeFn
+}
+
+func TestPriorityLevelIsolation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, true)()
+	// NOTE: disabling the feature should fail the test
+	_, loopbackConfig, closeFn := setup(t)
+	defer closeFn()
+
+	loopbackClient := clientset.NewForConfigOrDie(loopbackConfig)
+	noxu1Client := getClientFor(loopbackConfig, "noxu1")
+	noxu2Client := getClientFor(loopbackConfig, "noxu2")
+
+	queueLength := 50
+	concurrencyShares := 1
+
+	priorityLevelNoxu1, _, err := createPriorityLevelAndBindingFlowSchemaForUser(
+		loopbackClient, "noxu1", concurrencyShares, queueLength)
+	if err != nil {
+		t.Error(err)
+	}
+	priorityLevelNoxu2, _, err := createPriorityLevelAndBindingFlowSchemaForUser(
+		loopbackClient, "noxu2", concurrencyShares, queueLength)
+	if err != nil {
+		t.Error(err)
+	}
+
+	sharedConcurrency, err := getSharedConcurrencyOfPriorityLevel(loopbackClient)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if 1 != sharedConcurrency[priorityLevelNoxu1.Name] {
```

Comment:
During the interim period before the suggested config objects are added, the the concurrency limit for each of this test's two priority levels will be 4 (that is ceiling(10/3) ).  This test does not control the relative timing of the scrape here and the injection of the suggested config objects.  That makes this a potentially flaky test.

I can see two easy ways to fix this.  One would be to give each of the priority levels here such a large number of concurrency shares that they will each get a concurrency limit of 4 regardless of whether the suggested config objects have been injected.  The other is to make the server's total concurrency limit 3 and leave these priority levels with 1 share each; this would make each priority level's concurrency limit 1 regardless of whether the suggested config objects have been injected.

Reply:
I expected this code to just be done in a loop N times, and fail if it takes longer than that to reach the right value.

Reply:
>  This test does not control the relative timing of the scrape here and the injection of the suggested config objects

the function `framework.RunAMaster(masterConfig)` (in the `setup` function in this test file) only proceeds iff the apiserver's `/healthz` works, which already includes the healthiness of those registered post-start-hooks. so we can be sure that the suggested configs are injected before we checks the shared concurrency, am i missing anything?

https://github.com/kubernetes/kubernetes/blob/6747678a396ada2f4761ff104495ddcc3ef050b3/test/integration/framework/master_utils.go#L219-L228

Reply:
I am not sure what health means for a webhook, but I see no connection from succcess of injecting suggested config out to anything.

Reply:
The problem has been resolved by setting the server's total concurrency limit to 2.